### PR TITLE
EnC: Remove dependency on solution from AnalyzeDocumentAsync

### PR DESCRIPTION
**Pull request by tmat**
Required to implement https://github.com/dotnet/roslyn/issues/27735

### CODE REVIEW

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 1]
```diff
@@ -3361,20 +3348,17 @@ private bool TryMapParameter(ValueTuple<SyntaxNode, int> parameterKey, IReadOnly
...
-            [Out]ArrayBuilder<int> reverseCapturesMap,                 // {new capture index -> old capture index}
-            [Out]ArrayBuilder<SyntaxNode> newCapturesToClosureScopes,  // {new capture index -> new closure scope}
-            [Out]ArrayBuilder<SyntaxNode> oldCapturesToClosureScopes,  // {old capture index -> old closure scope}
-            [Out]List<RudeEditDiagnostic> diagnostics,
-            out bool hasErrors,
+            [Out]ArrayBuilder<int> reverseCapturesMap,
+            [Out]ArrayBuilder<SyntaxNode> newCapturesToClosureScopes,
+            [Out]ArrayBuilder<SyntaxNode> oldCapturesToClosureScopes,
+            [Out]List<RudeEditDiagnostic> diagnostics,                 // {new capture index -> old capture index}
+            out bool hasErrors,  // {new capture index -> new closure scope}
```
Comment (sharwell) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: ❓ Did you use a refactoring for this? Looks like it messed up these comments.

Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: Ye,s used Change signature refactoring.
@dpoeschl

Comment (dpoeschl) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: I thought we had a bug for this already, but I can't seem to find it. @brettfo and I had a couple features with this problem but never got a fix put together.

Comment (dpoeschl) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: Created https://github.com/dotnet/roslyn/issues/33801

Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: @dpoeschl Thanks!

**Comment by tmat** [comment]
@ivanbasov PTAL

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 2]
```diff
@@ -578,17 +573,17 @@ private static bool ReportFatalErrorAnalyzeDocumentAsync(ImmutableArray<ActiveSt
...
 
-            UpdateUneditedSpans(diagnostics, script.Match, oldText, newText, documentId, trackingService, oldActiveStatements, newActiveStatements, newExceptionRegions, updatedTrackingSpans);
+            UpdateUneditedSpans(diagnostics, script.Match, oldText, newText, documentId, trackingServiceOpt, oldActiveStatements, newActiveStatements, newExceptionRegions, updatedTrackingSpans);
 
             Debug.Assert(newActiveStatements.All(a => a != null));
 
             if (updatedTrackingSpans.Count > 0)
             {
-                trackingService.UpdateActiveStatementSpans(newText, updatedTrackingSpans);
+                trackingServiceOpt.UpdateActiveStatementSpans(newText, updatedTrackingSpans);
```
Comment (ivanbasov) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: >trackingServiceOpt [](start = 16, length = 18)

check for null?

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 3]
```diff
@@ -761,7 +756,7 @@ private static bool ReportFatalErrorAnalyzeDocumentAsync(ImmutableArray<ActiveSt
...
```
Comment (ivanbasov) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: >trackingServiceOpt [](start = 16, length = 18)

check for null

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 2 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: Not necessary. `updatedTrackingSpans` will be empty if there is no tracking service.

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs [hunk 4]
```diff
@@ -982,7 +977,7 @@ internal struct UpdatedMemberInfo
                 // We seed the method body matching algorithm with tracking spans (unless they were deleted)
                 // to get precise matching.
                 TextSpan trackedSpan = default;
-                bool isTracked = trackingService?.TryGetSpan(new ActiveStatementId(documentId, ordinal), newText, out trackedSpan) ?? false;
+                bool isTracked = trackingServiceOpt?.TryGetSpan(new ActiveStatementId(documentId, ordinal), newText, out trackedSpan) ?? false;
```
Comment (ivanbasov) [inline src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: >trackingServiceOpt? [](start = 33, length = 19)

nit: maybe for consistency, it is worth to use the same approach for all such checks: either this one "trackingServiceOpt?... ?? false" or the approach above "trackingServiceOpt != null && ... "

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 2 above)
Comment (ivanbasov) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: add a comment?

---
In reply to: [261778220](https://github.com/dotnet/roslyn/pull/33796#discussion_r261778220) [](ancestors = 261778220)

File: src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs (hunk 4 above)
Comment (tmat) [reply src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs]: I'll follow up.

**Pull request merged**